### ISSUE TITLE
[doc] No example or documentation of creating Connectors

### ISSUE THREAD

**Issue by gaurav21r** [strongloop/loopback#2050]
Greetings from https://github.com/rcorp We are in the process of migrating from Waterline / Sails.JS to loopback and we've found loopback fascinating so far!

We want to migrate our existing connectors (adapters in waterline speak) but we've found no documentation on how to do this apart from here: https://docs.strongloop.com/display/public/LB/Building+a+connector

This only covers adapters based off of **SQL**. and that too not all the required `methods` for the connector to work. Eg: what does the `exports.initialize()` supposed to do?  

It would be great if there is a solid guide for creating connectors, maybe like [this](https://github.com/balderdashy/sails-docs/blob/0.9/adapter-specification.md) or [this](https://github.com/balderdashy/waterline-docs/blob/master/adapters/adapters.md)? We would like to open source some of our connectors as community connectors as well as create a auto conversion mechanism to convert `Waterline Adapters` to `LoopBack Connectors`.

One other suggestion we'd like to give is to maybe add a yeoman generator (like [this](https://github.com/balderdashy/sails-generate-adapter) ) to scaffold the structure for a new connector? Maybe like the one `slc` does for `Models`?

**Comment by richardpringle** [comment]
@crandmck, I'll let you reply in terms of the doc comments.

@gaurav21r:
>One other suggestion we'd like to give is to maybe add a yeoman generator (like this ) to scaffold the structure for a new connector? Maybe like the one slc does for Models?

Interesting idea. Could you open a separate issue so a discussion can be formed around the feature request?

**Comment by gaurav21r** [comment]
@richardpringle Thanks for your response. Yep will do that! But I'll still request you to keep this issue open for the docs. I hope you & @crandmck had a look at the links. Even some simple Markdown documentation within https://github.com/strongloop/loopback-datasource-juggler/tree/master/docs is enough.

**Comment by gaurav21r** [comment]
I've prepared some *notes* based on my initial interaction with the docs and code digging -> 

I think the docs should cover these points and also, if someone can help with the clarifications? Please note, I'm trying to structure points as an **issue and not a discussion for the forum**. These are some things I think the docs should address for any developer who is interested in creating a custom `connector`.

## Instructions for Connectors

### A. Connectors inherit from `require(loopback-connector).Connector`

### B. Connector modules must export an `initialise()` function.

The initialise function gets data from the `dataSource` as parameters.
A `new` instance of the connector is to be setup in `dataSource.connector`.

Eg: from [loopback-connector-mongodb](https://github.com/strongloop/loopback-connector-mongodb/blob/master/lib/mongodb.js#L56)

```js
exports.initialize = function initializeDataSource(dataSource, callback) {
  if (!mongodb) {
    return;
  }

  var s = dataSource.settings;

  s.safe = (s.safe !== false);
  s.w = s.w || 1;
  s.url = s.url || generateMongoDBURL(s);
  dataSource.connector = new MongoDB(s, dataSource);
  dataSource.ObjectID = mongodb.ObjectID;

  if (callback) {
    dataSource.connector.connect(callback);
  }
};

```
 
### C. Methods defined in `Connector` should be overwritten.

#### Clarifications

1. What are the **required** methods to be overwittten?
2. What does calling the base class from within our Connector (`super()`) do? 


### D. CRUD Methods are to be derived from the DAO Facade

#### Clarifications

1. Should connectors inherit from https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/dao.js ?
2. Again, which methods should be **required** to implement?
3. If we implement a `create()` and an `update()` and don't specifiy an
`updateOrCreate()` will `dao.js` handle this for us automatically, through
https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/dao.js#L437 ?

  I strongly recommend that this should be the case. There is no need to create so many methods
  for our conncetors unless we want to run some special behaviour!

### E. Some methods have to implemented for relations to work.
https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/relation-definition.js

Eg: `findOrCreate()` is required for the `add()` function of `hasAndBelongsToMany` relations.
https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/relation-definition.js#L1028

#### Clarifications
1. Which method is required for Relations to work. Again, can inheriting DAO, help us?

**Comment by crandmck** [comment]
@raymondfeng PTAL ^

If the above looks good, I will add to the docs.  

Unfortunately, I doubt if I will be able to prioritize this work any time soon.  Perhaps @raymondfeng can give you some pointers.

**Comment by gaurav21r** [comment]
@crandmck I understand your constraint of resources and as I had mentioned if we could do #1922 it would be great! I didn't really write this in a **documentation** like language as such, more like rough notes.

@raymondfeng Could you please help me with those clarifications, docs and code have only made me come so far. I have to *pianfully* debug using **node-inspector** to actually see the control flow when executing test programs, to have even come this far!

@crandmck @raymondfeng @richardpringle May I ask why this can't be upped in priority? As mentioned, the docs are still better for **SQLConnectors** but with Strongloop's official connectors for 

  - [MySQL](https://github.com/strongloop/loopback-connector-mysql)
  - [Oracle](https://github.com/strongloop/loopback-connector-oracle)
  - [DB2](https://github.com/strongloop/loopback-connector-db2)
  - [Postgresql](https://github.com/strongloop/loopback-connector-postgresql)
  - [MSSQL](https://github.com/strongloop/loopback-connector-mssql)

and community connector for [SQLite](https://github.com/Synerzip/loopback-connector-sqlite)
you've pretty much covered **98% of SQL Land!**. I see very few Use Cases for user defined SQL Connectors.

Though, **REST** and **SOAP** do allow us to connect to a variety of services, there is still a HUGE need for us to develop connectors for unconventional sources especially those whose API do not follow the Swagger Template approach

**Comment by gaurav21r** [comment]
I believe **connectors** are at the heart of what brings **loopback** value. Any response for my queries / docs? I think it will greatly benefit us as well as the community. @raymondfeng @crandmck @altsang @bajtos @richardpringle @superkhau

**Comment by bajtos** [comment]
@gaurav21r I can't help you to get the documentation created sooner, but I can answer your questions. I think it's a good idea to keep all of these Q&A here in this issue, so that the person writing the doc page can find all information in a single place.

> D. CRUD Methods are to be derived from the DAO Facade
>     Should connectors inherit from https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/dao.js ?

No. `DataAccessObject` is mixed into model class when a model is attached to a data-source. By default, `juggler/lib/dao.js` is used. However, connectors may provide their own `DataAccessObject` - this would be the case for connectors that do not implement/support LoopBack's `PersistedModel` API. See https://github.com/strongloop/loopback-datasource-juggler/blob/9a76fb4c071cc4c811d0553980caae66bc8d2644/lib/datasource.js#L117

> Again, which methods should be required to implement?

If the connector is compatible with the built-in `DataAccessObject`, then it needs to implement methods called from `DataAccessObject` methods. You can find the required connector methods by running `grep "connector\.[a-zA-Z]*(" lib/dao.js`. Here is the distilled list:

 - connector.create
 - connector.updateOrCreate (optional, but see below)
 - connector.replaceOrCreate (a new feature - work in progress)
 - connector.findOrCreate (optional, but see below)
 - connector.buildNearFilter
 - connector.all
 - connector.destroyAll
 - connector.count
 - connector.save
 - connector.update
 - connector.destroy
 - connector.replaceById (a new feature - work in progress)
 - connector.updateAttributes

>  If we implement a create() and an update() and don't specifiy an
>  updateOrCreate() will dao.js handle this for us automatically, (...)
>  I strongly recommend that this should be the case. There is no need to create so many methods
>  for our conncetors unless we want to run some special behaviour!

Yes, juggler will handle `updateOrCreate` and `findOrCreate` if they are not provided by the connector. However, I am strongly recommending to implement these methods, because only the connector can provide a version that guarantees atomicity. The juggler shim makes two database/datasource calls (find and then update/create depending on the result), which creates a race condition. I.e. if you call `updateOrCreate({ id: 123, name: 'foo' })` twice at the same time (in parallel), the second call is very likely to fail as the second "find" will return before the first "create" was called, and thus the second call will attempt to create a record that already exists.

> Which method is required for Relations to work. Again, can inheriting DAO, help us?

I am not familiar with relations myself, but I would expect it is enough to use the default `DataAccessObject` and then implement the connector methods as described.

I would recommend you to look at the source code of the [memory](https://github.com/strongloop/loopback-datasource-juggler/blob/9a76fb4c071cc4c811d0553980caae66bc8d2644/lib/connectors/memory.js) connector. It does support relations and AFAICT it doesn't implement any special methods for that.

**Comment by superkhau** [comment]
@gaurav21r I too believe there is value for creating an example of how to create connectors. However, this will probably have to wait until we finish splitting out juggler into individual components. It'll be less confusing then. Let's add this one to the backlog.

**Comment by arbrebinaire** [comment]
@gaurav21r  "I believe connectors are at the heart of what brings loopback value."

The guy is so right.  He wrote that last Valentine's day.

Been working on a connector for JSONP resources (there is a lot of that around)  Quite an undertaking, since[ loopback-connector](https://github.com/strongloop/loopback-connector) (the "base" class) does not agree with the current docs, which is unprecise about what needs to be implemented.  The other implementations (mongo, mysql, etc) are not standardized enough to understand where data from foreign source is injected using the DAO.

Loopback is a fantastic tool: you thought of external data sources.  Let's focus on that for us "external" coders.

**Comment by crandmck** [comment]
OK, I tried to capture some of this information in https://docs.strongloop.com/display/APIC/Building+a+connector#Buildingaconnector-Notesforcreatingnon-databaseconnectors

Of course, this is just a bare start and still needs lots of work.  I'll continue revising it.  Please suggest any substantive changes or additions.

**Comment by arbrebinaire** [comment]
Hi,

Much appreciated!

Regards,

François

2016-06-29 18:57 GMT-04:00 Rand McKinney <notifications@github.com>:

**Comment by gaurav21r** [comment]
@crandmck THanks :+1: I'll go through it next week and leave any feedback here. Once again, I stress the fact that since loopback is good enough to be open source, the community can also shoulder some of the tasks of documentation!

**Comment by crandmck** [comment]
Great, thanks @gaurav21r ...  We are planning on moving the LB docs soon into an open-source format (similar to what we use for http://expressjs.com).  That will make it much easier for the community to contribute.

**Comment by activedecay** [comment]
I want to create a file-connector. I would like to use the filesystem as a DataSource, and I guess the documentation for lb3 should be consulted. Does one exist already? How do I create a DataSource/Connector for a filesystem using lb4?

[Building a connector](https://loopback.io/doc/en/lb3/Building-a-connector.html) This page mentions a "Storage connector" about half-way down the page. This is something I would like to use in lb4, but I have not been able to find it. Am I the first person to want/need a storage connector?

Am I able to use connectors from lb3 in lb4?

How do I add a connector to a lb4 project if the connector is not a node module that can be installed with npm install? I don't have an npm repository that I can upload private npm modules to -- is there a way to create a lb4 connector without using npm modules?
